### PR TITLE
[feature] Add CLEVR dataset

### LINKED ISSUE #94: [enhancement] Add support for CLEVR
## 🚀 Feature

CLEVR is one of the most of famous abstract reasoning datasets. 

## Motivation

Something similar to VQA but should return direct image so that models are able to handle them on their own.

### PR DESCRIPTION
**Pull request by apsdehal**
Fixes #94 
- Adds CLEVR dataset and builder classes
- If not present CLEVR's data will be automatically downloaded and
extracted
- This will also create vocabularies
- Also adds CLEVR base config

### CODE REVIEW

File: pythia/utils/general.py [hunk 1]
```diff
@@ -94,6 +94,9 @@ def download_file(url, output_dir=".", filename=""):
     filename = os.path.join(output_dir, filename)
     r = requests.get(url, stream=True)
 
+    if r.status_code != 200:
```
Comment (abhiskk) [inline pythia/utils/general.py]: Add enum for status

File: pythia/tasks/vqa/clevr/builder.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+@registry.register_builder("clevr")
+class CLEVRBuilder(BaseDatasetBuilder):
+    def __init__(self):
+        super().__init__("clevr")
+        self.writer = registry.get("writer")
+        self.dataset_class = CLEVRDataset
+
+    def _build(self, dataset_type, config):
+        download_folder = os.path.join(get_pythia_root(), config.data_root_dir)
+        download_folder = os.path.join(download_folder, config.data_folder)
```
Comment (mathfac) [inline pythia/tasks/vqa/clevr/builder.py]: Maybe
```suggestion
        download_folder = os.path.join(get_pythia_root(), config.data_root_dir, config.data_folder)
```

File: pythia/tasks/vqa/clevr/dataset.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+from pythia.common.registry import registry
+from pythia.common.sample import Sample
+from pythia.tasks.base_dataset import BaseDataset
+from pythia.utils.general import get_pythia_root
+from pythia.utils.text_utils import VocabFromText, tokenize
+from pythia.utils.distributed_utils import is_main_process, synchronize
+
+
+class CLEVRDataset(BaseDataset):
+    def __init__(self, dataset_type, config, data_folder=None, *args, **kwargs):
```
Comment (mathfac) [inline pythia/tasks/vqa/clevr/dataset.py]: Can we include `dataset_type` and `data_folder ` inside config?

File: pythia/tasks/vqa/clevr/dataset.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+from pythia.common.registry import registry
+from pythia.common.sample import Sample
+from pythia.tasks.base_dataset import BaseDataset
+from pythia.utils.general import get_pythia_root
+from pythia.utils.text_utils import VocabFromText, tokenize
+from pythia.utils.distributed_utils import is_main_process, synchronize
+
+
+class CLEVRDataset(BaseDataset):
```
Comment (mathfac) [inline pythia/tasks/vqa/clevr/dataset.py]: Maybe worth to add doc string.

File: pythia/tasks/vqa/clevr/dataset.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                self._build_vocab(self.questions, "question")
+                self._build_vocab(self.questions, "answer")
+            synchronize()
+
+    def __len__(self):
+        return len(self.questions)
+
+    def _get_vocab_path(self, attribute):
+        return os.path.join(
+            self._data_root_dir, "vocabs",
```
Comment (mathfac) [inline pythia/tasks/vqa/clevr/dataset.py]: create constants for all file templates and folders.

File: pythia/utils/text_utils.py [hunk 6]
```diff
@@ -88,20 +98,40 @@ def __init__(self, vocab_file, data_root_dir=None):
...
 
         self.word_list = load_str_list(vocab_file)
+        self._build()
+
+    def _build(self):
+        if self.UNK_TOKEN not in self.word_list:
+            self.word_list = [self.UNK_TOKEN] + self.word_list
+
         self.word2idx_dict = {w: n_w for n_w, w in enumerate(self.word_list)}
+        self.stoi = self.word2idx_dict
```
Comment (mathfac) [inline pythia/utils/text_utils.py]: Maybe add comment what `stoi` and `itos` stand for?

File: pythia/utils/text_utils.py [hunk 7]
```diff
@@ -120,6 +150,33 @@ def tokenize_and_index(self, sentence):
         return inds
 
 
+class VocabFromText(VocabDict):
+    DEFAULT_TOKENS = ["<pad>", "<unk>", "<s>", "</s>"]
```
Comment (mathfac) [inline pythia/utils/text_utils.py]: Maybe
```suggestion
    SPECIAL_TOKENS = [VocabDict.UNK_INDEX, VocabDict.PAD_TOKEN, VocabDict.SEN_START , VocabDict.SEN_END]
```

File: pythia/tasks/vqa/clevr/dataset.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            "{}_{}_vocab.txt".format(self._name, attribute)
+        )
+
+    def _build_vocab(self, questions, attribute):
+        # Don't build when not train
```
[the remark this answers is not in the document]
Comment (mathfac) [reply pythia/tasks/vqa/clevr/dataset.py]: `Don't build when not train` that's obvious to say from if statement. The question that comment should answer why  it shouldn't build for val/test?

File: pythia/tasks/vqa/vqa_task.py [hunk 9]
```diff
@@ -9,7 +9,7 @@ def __init__(self):
         super(VQATask, self).__init__("vqa")
 
     def _get_available_datasets(self):
-        return ["vqa2", "vizwiz", "textvqa", "vqa2_ocr"]
+        return ["vqa2", "vizwiz", "textvqa", "vqa2_ocr", "clevr"]
```
[the remark this answers is not in the document]
Comment (apsdehal) [reply pythia/tasks/vqa/vqa_task.py]: Delegating for another time. It is not possible to make the suggested change in current context.

File: pythia/tasks/vqa/clevr/dataset.py (hunk 3 above)
Comment (apsdehal) [reply pythia/tasks/vqa/clevr/dataset.py]: To keep the configuration same for train, val, test and avoid repetition, this needs to be added in this way. Though, these implementation choices depend on the developer and can be done differently too.

**Pull request merged**